ee/secureenclavesigner/secureenclavesigner_darwin.go

```diff
@@ -0,0 +1,295 @@
+//go:build darwin
+// +build darwin
+
+package secureenclavesigner
+
+import (
+	"context"
+	"crypto"
+	"crypto/ecdsa"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"os"
+	"os/user"
+	"strings"
+	"sync"
+
+	"github.com/kolide/krypto/pkg/echelper"
+	"github.com/kolide/launcher/ee/agent/types"
+	"github.com/kolide/launcher/ee/allowedcmd"
+	"github.com/kolide/launcher/ee/consoleuser"
+	"github.com/kolide/launcher/pkg/traces"
+)
+
+const (
+	CreateKeyCmd     = "create-key"
+	PublicEccDataKey = "publicEccData"
+)
+
+type opt func(*secureEnclaveSigner)
+
+type secureEnclaveSigner struct {
+	uidPubKeyMap         map[string]*ecdsa.PublicKey
+	pathToLauncherBinary string
+	store                types.GetterSetterDeleter
+	slogger              *slog.Logger
+	mux                  *sync.Mutex
+}
+
+func New(ctx context.Context, slogger *slog.Logger, store types.GetterSetterDeleter, opts ...opt) (*secureEnclaveSigner, error) {
+	ctx, span := traces.StartSpan(ctx)
+	defer span.End()
+
+	ses := &secureEnclaveSigner{
+		uidPubKeyMap: make(map[string]*ecdsa.PublicKey),
+		store:        store,
+		slogger:      slogger.With("component", "secureenclavesigner"),
+		mux:          &sync.Mutex{},
+	}
+
+	data, err := store.Get([]byte(PublicEccDataKey))
+	if err != nil {
+		traces.SetError(span, fmt.Errorf("getting public ecc data from store: %w", err))
+		return nil, fmt.Errorf("getting public ecc data from store: %w", err)
+	}
+
+	if data != nil {
+		if err := json.Unmarshal(data, ses); err != nil {
+			traces.SetError(span, fmt.Errorf("unmarshaling secure enclave signer: %w", err))
+			ses.slogger.Log(ctx, slog.LevelError,
+				"unable to unmarshal secure enclave signer, data may be corrupt, wiping",
+				"err", err,
+			)
+
+			if err := store.Delete([]byte(PublicEccDataKey)); err != nil {
```

Comment:
I think it's probably correct to delete the key if it's corrupt, but it feels a little odd to do it here in the secureenclavesigner.

Reply:
Yea it feels a little odd. I though it might be better than switching on errors in the calling function, deciding what to do, then calling new again. Also, since, without some big changes to the way `SetUpKeys` works, we have to let `secureEnclaveSIgner` handle the saving of new keys. So it kinda made sense to just let it handle the whole storage  layer.